asab/api/discovery.py

```diff
@@ -250,6 +257,10 @@ def get_data(item):
 			yield item, json.loads(item_data)
 
 
+	def _on_change(self, watched_event):
+		self.App.PubSub.publish_threadsafe("Discover.Rescan!")
```

Comment:
Comment this method to indicated clearly and loudly that this callback is happening on non-main thread.

Try more direct approach:

```
def _on_change(self, watched_event)
    self.App.Loop.call_soon_threadsafe(self._update_cache)
```